### PR TITLE
Add flags checks to BMI1 intrinsic lowering

### LINKED ISSUE #66709: JIT: Incorrect optimization of x & -x on x86
### Description

The following program gives an invalid result when optimizing on x86.

### Reproduction Steps

```csharp
// Generated by Fuzzlyn v1.5 on 2022-03-16 11:07:55
// Run on X86 Windows
// Seed: 13083630089409025722
// Reduced from 309.3 KiB to 0.2 KiB in 00:06:28
// Debug: Outputs 1
// Release: Outputs 4294967297
public class Program
{
    public static long s_61 = -1;
    public static void Main()
    {
        s_61 &= -s_61;
        System.Console.WriteLine(s_61);
    }
}
```

### Expected behavior

Same result with and without optimizations.

### Actual behavior

Different result.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

Probably some interaction between long decomposition and the new optimization added in #66193? cc @Wraith2 @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by Wraith2**
Fixes https://github.com/dotnet/runtime/issues/66709

On x86 longs are decomposed into two integer nodes. This bug was caused by one of those integer nodes being recognised as part of the `blsi` pattern, `x & -x` and transformed without the second node being included. On x86 the 64 bit version of the intrinsic is not available so the lowering of a 64 bit operation to the intrinsic is not possible.

This PR adds in flags checks on all nodes being removed to see if they have side effect and if they have the lowering is not attempted. 

I have added the same checks to `blsr` and `andn` even though I have confirmed that they do not suffer from the same bug. This seemed prudent to me in case future jit changes expose a situation where it does become possible for those lowering to be affected.

/cc @dotnet/jit-contrib

### CODE REVIEW

**Comment by jakobbotsch** [comment]
Can you add the regression test?

**Review by AndyAyersMS — COMMENTED** [review]
Can you also add the test case that uncovered this?

Feel free to add it to your existing tests, if that works out.

**Comment by Wraith2** [comment]
I've updated the bmi1 tests to have 64 bit overloads and verified that they fail on current main as expected.

**Review by AndyAyersMS — APPROVED** [review]

File: src/coreclr/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -3789,6 +3789,13 @@ GenTree* Lowering::TryLowerAndOpToResetLowestSetBit(GenTreeOp* andNode)
         return nullptr;
     }
 
+    // prevent decomposed 64 integer node parts in x86 from being recognised
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/coreclr/jit/lowerxarch.cpp:3792]: > We're checking the general "does this node have a side-effect", it so happens that for these nodes the only possible side effect is setting the flags.

Ok. but the thing I'm setting is called `GTF_SET_FLAGS` which has no obvious meaning. The comment explains _why_ I'm checking it which is important for anyone reading the code. If you've been deep inside the JIT codebase for years you've probably picked up the meaning from somewhere or had it explained but to a casual viewer this is nonsensical. 

In a lot of other places in the codebase there are checks but no explanation of why they're being made so I'd like to keep a comment which explains the why not only the what. I'm open to rephrasing but I feel there needs to be some clear explanation for the code being present.

File: src/coreclr/jit/lowerxarch.cpp [hunk 2]
```diff
@@ -3789,6 +3789,13 @@ GenTree* Lowering::TryLowerAndOpToResetLowestSetBit(GenTreeOp* andNode)
...
+    if (((addOp1->gtFlags & GTF_SET_FLAGS) == GTF_SET_FLAGS) || ((addOp2->gtFlags & GTF_SET_FLAGS) == GTF_SET_FLAGS) ||
+        ((op2->gtFlags & GTF_SET_FLAGS) == GTF_SET_FLAGS) || ((andNode->gtFlags & GTF_SET_FLAGS) == GTF_SET_FLAGS))
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/coreclr/jit/lowerxarch.cpp:3794]: Ok, so how would you programmatically work out which nodes could have side effects?

File: src/coreclr/jit/lowerxarch.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (SingleAccretion) [reply src/coreclr/jit/lowerxarch.cpp:3792]: > I'm open to rephrasing but I feel there needs to be some clear explanation for the code being present.

Sure, I'd go with `Take care not to remove nodes that must set flags.`

File: src/coreclr/jit/lowerxarch.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (SingleAccretion) [reply src/coreclr/jit/lowerxarch.cpp:3794]: There is not a way for this right now (one could be added, say `OperCanSetFlags`), but I simply cannot imagine a world where `LCL_VAR`s  would return `true` from it, there is a lot of code that relies on local leafs being side-effect free.

File: src/coreclr/jit/lowerxarch.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/coreclr/jit/lowerxarch.cpp:3792]: That only explains what it's doing not why. The meaning of "setting flags" is entirely missing. If it were named something like `GTF_HAS_SIDE_EFFECTS` it'd be different but the name `GTF_SET_FLAGS` is about as descriptive as calling it `GTF_BIT7`.

How about: `// take care not to remove nodes which may have side effects indicated by GTF_SET_FLAGS, for example x86 decomposed 32 bit values.` That explains how the check code correlates to the problem that is being solved and links the name of the enum value to its purpose.

[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/lowerxarch.cpp:3792]: "Subsequent nodes may rely on CPU flags set by this node in which case we cannot remove it"?

[the remark this answers is not in the document]
Comment (Wraith2) [reply src/coreclr/jit/lowerxarch.cpp]: Done.

File: src/coreclr/jit/lowerxarch.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/coreclr/jit/lowerxarch.cpp]: removed that node from the checks and updated them all to do !=0 comparisons.

**Review by SingleAccretion — APPROVED** [review]

**Review by jakobbotsch — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Failure is #66571.

**Pull request merged**